pkg/providers/gcp/provider_postgres.go

```diff
@@ -3,63 +3,230 @@ package gcp
-		return nil, types.StatusMessage(errMsg), fmt.Errorf("%s: %w", errMsg, err)
+		return nil, croType.StatusMessage(errMsg), fmt.Errorf("%s: %w", errMsg, err)
+	}
+
+	if err := resources.CreateFinalizer(ctx, pp.Client, p, DefaultFinalizer); err != nil {
+		return nil, "failed to set finalizer", err
 	}
+
 	// TODO implement me
 	return nil, "", nil
 }
 
-func (pp PostgresProvider) DeletePostgres(ctx context.Context, p *v1alpha1.Postgres) (types.StatusMessage, error) {
-	_, err := pp.CredentialManager.ReconcileProviderCredentials(ctx, p.Namespace)
+func (pp *PostgresProvider) DeletePostgres(ctx context.Context, p *v1alpha1.Postgres) (croType.StatusMessage, error) {
+	logger := pp.Logger.WithField("action", "DeletePostgres")
+	logger.Infof("reconciling postgres %s", p.Name)
+
+	// set postgres deletion timestamp metric
+	pp.setPostgresDeletionTimestampMetric(ctx, p)
+
+	// get provider gcp creds to access the postgres instance
+	creds, err := pp.CredentialManager.ReconcileProviderCredentials(ctx, p.Namespace)
 	if err != nil {
 		errMsg := fmt.Sprintf("failed to reconcile gcp postgres provider credentials for postgres instance %s", p.Name)
-		return types.StatusMessage(errMsg), fmt.Errorf("%s: %w", errMsg, err)
+		return croType.StatusMessage(errMsg), fmt.Errorf("%s: %w", errMsg, err)
 	}
-	// TODO implement me
-	return "", nil
+
+	// build cloudSQL service
+	sqladminService, err := sqladmin.NewService(ctx, option.WithCredentialsJSON(creds.ServiceAccountJson))
+	if err != nil {
+		return "error building cloudSQL admin service", err
+	}
+	sqlClient := &sqlClient{
+		sqlAdminService: sqladminService,
+	}
+	return pp.deleteCloudSQLInstance(ctx, sqlClient, p)
 }
 
-var _ providers.PostgresProvider = (*PostgresProvider)(nil)
+func (pp *PostgresProvider) deleteCloudSQLInstance(ctx context.Context, sqladminService SQLAdminService, p *v1alpha1.Postgres) (croType.StatusMessage, error) {
+	logger := pp.Logger.WithField("action", "deleteCloudSQLInstance")
+	// get cloudSQL instance
+	instances, err := getCloudSQLInstances(sqladminService)
+	if err != nil {
+		return "cannot retrieve sql instances from gcp", err
+	}
+	logrus.Info("listed sql instances from gcp")
+
+	instanceName := annotations.Get(p, ResourceIdentifierAnnotation)
+
+	if instanceName == "" {
+		msg := "unable to find instance name from annotation"
+		return croType.StatusMessage(msg), fmt.Errorf(msg)
+	}
+
+	// check if instance exists
+	var foundInstance *sqladmin.DatabaseInstance
+	for _, instance := range instances {
+		if instance.Name == instanceName {
+			logrus.Infof("found matching instance by name: %s", instanceName)
+			foundInstance = instance
+			break
+		}
+	}
+	// check instance state
+	if foundInstance != nil {
+		if foundInstance.State != "RUNNABLE" {
+			statusMessage := fmt.Sprintf("delete detected, DeletePostgres() is in progress, current cloudSQL status is %s", foundInstance.State)
+			pp.Logger.Info(statusMessage)
+			return croType.StatusMessage(statusMessage), nil
+		}
```

Comment:
Why can't we proceed with the instance deletion if the [instance state](https://pkg.go.dev/google.golang.org/api/sqladmin/v1beta4@v0.66.0#DatabaseInstance.State) is not `RUNNABLE`? Additionally, I think the status message is a bit misleading as the instance is not guaranteed to be in the process of being deleted yet.

Reply:
The idea here is that if the state is not "RUNNABLE" then a delete is already in progress, if it is "RUNNABLE" then go ahead and delete it now. This is the way it is done for AWS RDS. I can update the status msg to be more specific.

Reply:
What I had in mind is the following scenario:

If a user creates Postgres CR and CRO starts provisioning the cloud resource in GCP, the logic for deletion will not be triggered until the cloud resource is finished provisioning (`RUNNABLE`). You should probably changing this if statement to check for state `PENDING_DELETE`. These are the possible states:
```
"SQL_INSTANCE_STATE_UNSPECIFIED" - The state of the instance is unknown.
"RUNNABLE" - The instance is running, or has been stopped by owner.
"SUSPENDED" - The instance is not available, for example due to problems with billing.
"PENDING_DELETE" - The instance is being deleted.
"PENDING_CREATE" - The instance is being created.
"MAINTENANCE" - The instance is down for maintenance.
"FAILED" - The creation of the instance failed or a fatal error occurred during maintenance.
```

@laurafitzgerald, what's your input on this change?

Reply:
That makes sense. Happy to change it but will wait for Laura to confirm. Thanks.

Reply:
If the other states allow for a successful delete request then let's update the check, It makes sense to me to do a delete on any state that is !PENDING_DELETE as long as those work okay.